hal/src/nRF52840/usart_hal.cpp

```diff
@@ -542,21 +587,25 @@ class Usart {
         }
     }
 
-    void startReceiver(bool flush = false) {
-        if (receiving_ >= MAX_SCHEDULED_RECEIVALS) {
-            return;
+    size_t prepareNextRingBuffer(bool flush = false) {
+        // Updates current size
+        particle::services::RingBuffer<uint8_t>* buffer = nullptr;
+        if (usingAltRxBuffer_) {
+            buffer = &rxBuffer_;
+        } else {
+            buffer = &rxBufferAlt_;
         }
 
-        // Updates current size
-        rxBuffer_.acquireBegin();
+        // To keep the received data contiguous
+        if (buffer->data()) {
+            return 0;
+        }
```

Comment:
Somewhere decided we needed to prepare the next buffer but it has unconsumed bytes so rejected. Is this a surface for the original problem to present again? Or something else bad?

Reply:
If we don't read out received data in time, then yes. But this is a common issue even if we use a super large buffer if user don't read out the data in time. We will configure the another buffer in the `read()` method as long as the another buffer is empty.

Reply:
So it will continue to safely fill the current buffer without issue so long as it has room? Why is there a need for two buffers then?

Reply:
>it will continue to safely fill the current buffer without issue so long as it has room?

Yes. The secondary buffer aims to make sure when the ENDRX event is assert hardware can be immediately ready to receive the incoming data, even if the ENDRX event is not handled by ISR. We currently have to wait until the ENDRX event is handled by software and then start to receive the incoming data. There is a gap between when the ENDRX event is asserted and when the receiver is ready to receive data. This gap causes data lost.